### PR TITLE
Advanced Fields in URL

### LINKED ISSUE #348: facets, advanced search, etc, in URL
- [x] facet options - Issue #391, PR #428
- [x] date range filter PR #428 
- [ ] advanced search (on this ticket)
- [x] sort -- in #364

- [x] ~~anything else not covered which we pass to elasticsearch.~~

### PR DESCRIPTION
**Pull request by adamcee**
Closes #348 

Advanced fields in URL and all functionality should be working. E2E tests added. Unit tests updated as needed. For routing, set `inherit=false` so that for state transitions query params must be explicitly set each time.

-------------------------------
PR Review
- [ ] @joshuago78 
- [x] @jfranco-gri

### CODE REVIEW

File: src/client/app/core/search-service.js [hunk 1]
```diff
@@ -30,21 +30,23 @@
       },
       opts: getDefaultOptsObj(),
       facetCategoriesList: ['creator', 'grp_contributor', 'language', 'subject'],
+      advFieldsList: ['creator', 'grp_contributor', 'language', 'subject', 'title', 'date'],
```
Comment (jfranco-gri) [inline src/client/app/core/search-service.js]: Why don't you use the values in the ADVANCED_SEARCH constant to get the advanced fields?

File: src/client/app/core/search-service.js [hunk 2]
```diff
@@ -99,16 +102,27 @@
...
 
+      // advanced fields
+       this.advFieldsList.forEach(function(fieldName){
+         queryParams['adv_' + fieldName] = [];
+       });
+      this.opts.advancedFields.forEach(function(field){
+        var paramName = 'adv_' + field.field.name;
+        queryParams[paramName].push(field.term);
+      });
+
```
Comment (jfranco-gri) [inline src/client/app/core/search-service.js]: it seems like field.name should be 'adv_title' or 'adv_date' etc since it seems you have to build that each time you use it.

File: src/client/app/core/search-service.js [hunk 3]
```diff
@@ -392,7 +425,8 @@
      * based on query params and an Elasticsearch query executed.
      */
     function transitionStateAndSearch(){
-      $state.go('searchResults', this.buildQueryParams(), {reload: true});
+      console.log('SearchService::transitionStateAndSearch()');
```
Comment (jfranco-gri) [inline src/client/app/core/search-service.js]: I think this console.log can be removed

**Comment by jfranco-gri** [comment]
SearchService is getting pretty big and some of the functions are currently unused or redundant.  From what I can tell, these functions are not being used at all:

 `newSearch`
`updateSearch`
`parseFacetsToObj`
`clearFacetsIn`
`updateDate`

`executeSearch` is used, but much of the code for setting default values and checking searchService's opts is already in `updateOpts` and `resetOps` which are called before `executeSearch`.

File: src/client/app/core/search-service.js (hunk 1 above)
Comment (adamcee) [inline src/client/app/core/search-service.js]: I thought about it. But since we already have to type out advanced field names in the URL for the `SearchResults` state and again in the resolve section for said state, I thought "well, the code already can't be DRY.' In that case an array of strings is simpler to read than `Object.keys` or iterating over ADVANCED_FIELD properties to get `someProperty.name` ....

Comment (jfranco-gri) [inline src/client/app/core/search-service.js]: Well, this is an extra list to maintain if we decide to change those fields (we've already changed them a few times)

File: src/client/app/core/search-service.js (hunk 2 above)
Comment (adamcee) [inline src/client/app/core/search-service.js]: I agree. I was originally thinking it would function like `FACET.someFacet.name` but its really only being used for query parameters now ... and I am going to rename it to something like `field.queryParam`...

Comment (jfranco-gri) [inline src/client/app/core/search-service.js]: yeah naming it field.queryParam makes sense

File: src/client/app/core/search-service.js (hunk 1 above)
Comment (adamcee) [inline src/client/app/core/search-service.js]: True. I will change and see how the code looks.

File: src/client/app/core/search-service.js (hunk 3 above)
Comment (adamcee) [inline src/client/app/core/search-service.js]: Done

File: src/client/app/app.config.routing.js [hunk 4]
```diff
@@ -61,7 +72,21 @@
...
+                  var advField = ss.buildAdvancedField(field, value);
+                  if(advField){
+                    searchOpts.advancedFields.push(advField);
+                  }
+                });
+              }
+            });
+
             ss.updateOpts(searchOpts);
+
```
Comment (jfranco-gri) [inline src/client/app/app.config.routing.js]: For me, I don't think using `updateOpts` is the best to do here anymore. The primary purpose of `updateOpts` is to merge existing ops with new opts.  But here we are creating a brand new opts object and don't intend to use any previous options (as evidenced by the previous call to `resetOpts`). To me it feels like a brittle way to set the options here. I would rather we directly set ss.opts, such as `ss.opts = searchOpts` (or create and call a function that does that).  Then we wouldn't have to call `resetOpts` either.  Calling `resetOpts` earlier also feels weird because by calling it, we construct and set a default object that we have no intention of using and we then immediately construct another object with default properties set.

Comment (adamcee) [inline src/client/app/app.config.routing.js]: I agree about the confusion of the function names vs how they are being used, and the design/concept stuff you point out. But also .... 

The one thing calling `ss.resetOpts()` followed by `ss.updateOpts(searchOpts)` does do, though, is programmatically ensure that `SearchService.options` will always have the same object structure.

The `resetOpts()` call builds default structure. `searchOptions` in the resolve has some but not all default settings. Plus this may change.

Also, `resetOpts()` and `updateOpts()` are both tested code. Though we actually should write some sort of unit test for the logic that's in the resolve now (it's covered by E2E but not explicitly unit tested).

Comment (adamcee) [inline src/client/app/app.config.routing.js]: With regards to the unused SearchService functions, I agree the file is getting big and that it might be worth breaking it up. I definitely think we should get rid of `newSearch` and `updateSearch`, those have been unused for quite a while.

`parseFacetsToObj` and `clearFacetsIn` I would prefer to keep for now since the code is written and tested. If we don't use them in say a month then yeah we should definitely get rid of them but I think they may be useful to have.

`updateDate` I am hoping we will use fairly soon as a step towards updating `SearchService.opts` with individual functions for each option (like we'd talked about offline) as opposed to using `updateOpts`. 

But if you have strong feelings about getting rid of those functions we can do that, though maybe in another PR ....

**Comment by adamcee** [comment]
@jfranco-gri All issues should be addressed. I had some comments around some of the changes you suggested or issues raised. Let me know if you think any of them should be tackled in this PR.

**Comment by adamcee** [comment]
*(Reposting comments as they were on an outdated diff and got hidden)*

I agree about the confusion of the function names vs how they are being used, and the design/concept stuff you point out. But also ....

The searchOptionsvariable in the resolve has some but not all default opts settings (for example it doesn't have the sort default setting).

Calling ss.resetOpts() followed by ss.updateOpts(searchOpts) programmatically ensure that SearchService.options will always have the same object structure and correct defaults. It also anticipates change, as long as updateOpts() works as expected it should gracefully handle any mistakes made in building searchOptions in the resolve.

Also, resetOpts() and updateOpts() are both tested code. Though we actually should write some sort of unit test for the logic that's in the resolve now (it's covered by E2E but not explicitly unit tested).

I think making changes to address the issues you raise is worth doing, not sure though if this PR is the right place to do it. Curious what you think about the points I raised.

----------
With regards to the unused SearchService functions, I agree the file is getting big and that it might be worth breaking it up. I definitely think we should get rid of newSearch and updateSearch, those have been unused for quite a while.

parseFacetsToObj and clearFacetsIn I would prefer to keep for now since the code is written and tested. If we don't use them in say a month then yeah we should definitely get rid of them but I think they may be useful to have.

updateDate I am hoping we will use fairly soon as a step towards updating SearchService.opts with individual functions for each option (like we'd talked about offline) as opposed to using updateOpts.

But if you have strong feelings about getting rid of those functions we can do that, though maybe in another PR ....

**Comment by adamcee** [comment]
@jfranco-gri I played around with altering the resolve as you suggested, by doing `searchOpts = ss.getDefaultOptsObj()` and then assigning params to it. But if we don't use `ss.updateOpts()` we have to hard-code defaults for fields not in the URL into the resolve, which doesn't seem like the right way to do it since we would end up re-producing much of the logic in `updateOpts()`.

**Comment by adamcee** [comment]
@jfranco-gri Changes made as per our discussion and planning offline. Not using `SearchService.opts` at all in the resolve now, building a new opts object and then setting `ss.opts = opts`.  PR should pass all tests and all issues should be addressed.

**Comment by adamcee** [comment]
@joshuago78 Merge conflict resolved and all tests pass.

**Comment by joshuago78** [comment]
I'm getting over a dozen failing tests on this branch, most of them relating to Facet Sidebar. @susanley can you run tests on this branch and confirm?

**Comment by susanley** [comment]
@joshuago78 Sure, will run tests now.

**Comment by adamcee** [comment]
@joshuago78 Are you getting failures in Firefox or Chrome? I have been finding Chrome still fails for me and just use FF for E2E tests.

**Comment by joshuago78** [comment]
I am using Chrome, but these tests passed on all the other branches I have looked at today.

**Comment by susanley** [comment]
@joshuago78 @adamcee All tests pass for me in FF, one failure in Chrome. ```
Failures:
1) Facet Modal should deactivate previously checked facets correctly
  Message:
    Failed: unknown error: Element is not clickable at point (582, 146). Other element would receive the click: <li class="filter-checkboxes ng-scope" ng-repeat="facet in currentFacets">...</li>
      (Session info: chrome=50.0.2661.102)
      (Driver info: chromedriver=2.21.371459 (36d3d07f660ff2bc1bf28a75d1cdabed0983e7c4),platform=Mac OS X 10.10.5 x86_64) (WARNING: The server did not provide any stacktrace information)
    Command duration or timeout: 29 milliseconds
    Build info: version: '2.51.0', revision: '1af067d', time: '2016-02-05 19:15:17'
    System info: host: 'GT26738', ip: '153.10.131.141', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.10.5', java.version: '1.8.0_73'
    Driver info: org.openqa.selenium.chrome.ChromeDriver
    Capabilities [{applicationCacheEnabled=false, rotatable=false, mobileEmulationEnabled=false, chrome={chromedriverVersion=2.21.371459 (36d3d07f660ff2bc1bf28a75d1cdabed0983e7c4), userDataDir=/var/folders/0q/5g_d2rln60zbbqh8jzykcsj00000gn/T/.org.chromium.Chromium.711yRB}, takesHeapSnapshot=true, databaseEnabled=false, handlesAlerts=true, hasTouchScreen=false, version=50.0.2661.102, platform=MAC, browserConnectionEnabled=false, nativeEvents=true, acceptSslCerts=true, locationContextEnabled=true, webStorageEnabled=true, browserName=chrome, takesScreenshot=true, javascriptEnabled=true, cssSelectorsEnabled=true}]
    Session ID: bd51735f7a16cecc6d773a44cf3dbc5d
```

**Comment by adamcee** [comment]
**TL;DR**
Chrome E2E tests fail for me on develop and this branch. Firefox tests past for me on both branches. I spot-checked 3 Chrome test failures I had on this branch and they all work fine manually. If there are specific tests from the failures you saw on your machine  I should look at let me know. 

However, because of this and because tests pass for Susie (and I assume Jenny) in Firefox I have a high degree of confidence that this code is behaving as expected. I think our varying E2E results are because of timing issues with the tests across machines/environments and not the behavior of the app itself. I provide evidence for this below.

Josh I don't have an explanation why other tests were passing for you. I would be curious if you ran the failing tests for this branch one file at a time if you were able to reproduce the same errors.

Setting up Travis or another third-party service may take care of our E2E issue and is probably a good next step after going live.

I think until we solve this we can be fairly confident if two or three developers get tests to pass and if we also spot check some failures and they pass manually, that the code is behaving as expected.


--------------------
@joshuago78 I ran E2E tests in Firefox and Chrome, both on Develop and this PR branch. In both cases many Chrome tests failed while no Firefox tests did. 

I then looked at the Chrome failures on this PR branch and tested the first three manually.  All three behaviors work fine.

There was one test failure in Chrome in particular which clearly shows that my Chrome failures are because of timing issues. This is also something we'd theorized before. Here it is:

> 3) Search Results should gray out facet chips with zero limiting results
[chrome #1]   Message:
[chrome #1]     Failed: element not visible

> [chrome #1]     Error
[chrome #1]         at Object.create.toggleFacetOption.value (/Users/acahan/projects/portal/test/page_objects/results.page.js:205:46)

The test fails because a facet option element was not visible. When you load and look at the same page manually the facet element is there. The test is running too fast and not waiting for angular/the browser to load.

When conducted manually this test works.

From this, and because tests in Firefox also passed for Susie, I have a high degree of confidence that the code for this PR branch is behaving correctly.

Josh I suspect that if Chrome failures are timing-related, the addition of new E2E tests could cause failures b/c timing between tests is now changed. It may be that if you run individual spec files for this branch some will pass - or, that doing so, or re-running the whole spec several times, will produce *different* failures.

Because we clearly have some issues with reproducing E2E test results across different browsers and machines I propose that for cases like this in the future we rely on a quorum of either 2 or 3 developers - if 2 or 3 people can get tests to past in at least one browser, then we can assume that the tests pass. We could also supplement this with manual spot-checking as I have of some of the failing tests, but I do not think this would be practical to do for the entire set of failures.

Longer-term we should figure out how to make our E2E tests more reliable and/or see how other people have dealt with this issue.

**Comment by adamcee** [comment]
@joshuago78  Just following up on this.
